File: src/util/Sort.java
package util;

import java.util.Comparator;

import javafx.scene.control.TableColumn;
import javafx.scene.control.TreeTableColumn;

/**
 * Sort type.
 *
 * @author Martin Polakovic
 */
public enum Sort {
	/** From minimal to maximal element. */
	ASCENDING,
	/** From maximal to minimal element. */
	DESCENDING,
	/** No sort. Order remains as is. */
	NONE;

	/**
	 * Modifies a comparator to uphold the sorting type. If this is descending, new reverse
	 * comparator is returned, otherwise no new comparator is built. By definition every comparator
	 * is ascending, hence the comparator itself is returned when this is ascending. When this is
	 * none, no order comparator (static instance) is returned.
	 * <p/>
	 * Applying this method multiple times will not produce any new comparator instances. {@link #ASCENDING}
	 * and {@link #NONE} will consistently return the same instance and {@link #DESCENDING} will flip between the reverse
	 * and original instance of the comparator, i.e., reverse of a reverse will be the same comparator
	 * object/instance. Formally:
	 * <p/>
	 * c==DESCENDING.cmp(DESCENDING.cmp(c)) is always true.
	 * <p/>
	 * Some code uses null comparator as a no comparator, hence this method accepts null. In such
	 * case, null is always returned (no order has no reverse order).
	 *
	 * @return null if c null, c if ascending, reverse to c if descending or no order
	 * comparator (which always returns 0) when none.
	 */
	@SuppressWarnings("unchecked")
	public <T> Comparator<T> cmp(Comparator<T> c) {
		if (c==null) return null;
		switch (this) {
			case ASCENDING: return c;
			// note the java implementation of ReversedComparator causes multiple calls to reversed()
			// order to preserve the reference of the original comparator, i.e., calling reversed()
			// twice will return the original object. This can be very important.
			case DESCENDING: return c.reversed();
			case NONE: return (Comparator) util.functional.Util.SAME;
			default: throw new SwitchException(this);
		}
	}

	/** @return a corresponding sort for {@link TableColumn.SortType}. */
	public static Sort of(TableColumn.SortType sort) {
		switch (sort) {
			case ASCENDING  : return Sort.ASCENDING;
			case DESCENDING : return Sort.DESCENDING;
			default : throw new SwitchException(sort);
		}
	}

	/** @return a corresponding sort for {@link TreeTableColumn.SortType}. */
	public static Sort of(TreeTableColumn.SortType sort) {
		switch (sort) {
			case ASCENDING  : return Sort.ASCENDING;
			case DESCENDING : return Sort.DESCENDING;
			default : throw new SwitchException(sort);
		}
	}

	/** @return sort of the given column based on its {@link TableColumn.SortType}. */
	public static Sort of(TableColumn<?,?> column) {
		return of(column.getSortType());
	}

	/** @return sort of the given column based on its {@link TreeTableColumn.SortType}. */
	public static Sort of(TreeTableColumn<?,?> column) {
		return of(column.getSortType());
	}
}

File: src/util/Sort.kt
package util

import javafx.scene.control.TableColumn
import javafx.scene.control.TreeTableColumn
import java.util.*


/** @see Sort.of */
fun <T> Comparator<T>?.inSort(sort: Sort) = sort.of(this)

/** Sort type. */
enum class Sort {
    /** From minimal to maximal element. */
    ASCENDING,
    /** From maximal to minimal element. */
    DESCENDING,
    /** No sort. Order remains as is. */
    NONE;

    /**
     * Modifies the comparator to uphold this sorting type. If this is descending, new reverse comparator is returned,
     * otherwise no new comparator is built. By definition every comparator is ascending, hence the comparator itself is
     * returned when this is ascending. When this is none, no order comparator (static instance) is returned.
     *
     * Applying this method multiple times will not produce any new comparator instances. [.ASCENDING] and [.NONE]
     * consistently return the same instance and [.DESCENDING] will flip between the reverse and
     * original instance of the comparator, i.e., reverse of a reverse will be the same comparator object/instance.
     * Formally:
     *
     * c==c.cmp(DESCENDING).cmp(DESCENDING) is always true.
     *
     * Some code uses null comparator as a no comparator, hence this method accepts null. In such case, null is always
     * returned (no order has no reverse order).
     *
     * @return null if c null, c if ascending, reverse to c if descending or no order comparator (which always returns
     * 0) when none.
     */
    fun <T> of(c: Comparator<T>?): Comparator<T>? {
        if (c==null) return null
        return when (this) {
            ASCENDING -> c
            // We need to preserve reference
            // Rely on java impl.: ReversedComparator preserves the reference of the original comparator
            DESCENDING -> c.reversed()
            NONE -> util.functional.Util.SAME as Comparator<T>
        }
    }

    companion object {

        /** @return a corresponding sort for [TableColumn.SortType] */
        @JvmStatic fun of(sort: TableColumn.SortType) = when (sort) {
            TableColumn.SortType.ASCENDING -> Sort.ASCENDING
            TableColumn.SortType.DESCENDING -> Sort.DESCENDING
            else -> throw SwitchException(sort)
        }

        /** @return a corresponding sort for [TreeTableColumn.SortType] */
        @JvmStatic fun of(sort: TreeTableColumn.SortType) = when (sort) {
            TreeTableColumn.SortType.ASCENDING -> Sort.ASCENDING
            TreeTableColumn.SortType.DESCENDING -> Sort.DESCENDING
            else -> throw SwitchException(sort)
        }

        /** @return sort of the given column based on its [TableColumn.SortType] */
        @JvmStatic fun of(column: TableColumn<*,*>) = of(column.sortType)

        /** @return sort of the given column based on its [TreeTableColumn.SortType] */
        @JvmStatic fun of(column: TreeTableColumn<*, *>) = of(column.sortType)

    }
}